README.rst

```diff
@@ -1962,6 +1968,344 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
+the representation that makes for easy querying. As a result, users' queries frequently
+include complex and repetitive query structures that work around the database's chosen data model.
+
+The compiler's macro system empowers users *reshaping* their data's structure to fit
+their mental model, minimizing query complexity and repetitiveness without requiring changes
+to the shape of the data in the underlying data systems. The compiler achieves this by allowing
+users to define **macros** -- type-safe rules for programmatic query rewriting
+that transform user-provided queries on the *desired* data model into
+queries on the *actual* data model in the underlying data systems.
+
+When macros are defined, the compiler loads them into a macro registry -- a data structure used to
+track all currently available macros, the resulting GraphQL schema (accounting for macros), and
+any additional metadata needed by the compiler. The compiler then leverages this registry
+to expand queries that rely on macros, rewriting them into equivalent queries that do not contain
+any macros and therefore reflect the actual underlying data model. This makes them somewhat similar
+to SQL's idea of non-materialized views, though there are some key differences:
+
+- SQL views require database access and special permissions; databases are
+  completely oblivious to the use of macros since by the time the database gets the query,
+  all macro uses have been already expanded.
+
+- Macros can be stored and expanded client-side, so different users that query the same system may
+  define their own personal macros which are not shared with other users or the server that executes
+  the users' GraphQL queries. This is generally not achievable with SQL.
+
+- Since macro expansion does not interact in any way with the underlying data system, it works
+  seamlessly with all databases and even on schemas stitched together from multiple databases.
+  In contrast, not all databases support SQL-like :code:`VIEW` functionality.
+
+For now, the compiler supports one type of macro, **macro edges**. We hope to add more types
+of macros in the future.
+
+Macro registry
+~~~~~~~~~~~~~~
+
+The macro registry is where the definitions of all currently defined macros are stored,
+together with the resulting GraphQL schema they form, as well as any associated metadata
+that the compiler's macro system may need to expand any macros encountered in a query.
+
+To create a macro registry object for a given GraphQL schema, use the :code:`create_macro_registry`
+function:
+
+.. code:: python
+
+    from graphql_compiler.macros import create_macro_registry
+
+    macro_registry = create_macro_registry(your_graphql_schema_object)
+
+To retrieve the GraphQL schema object with all its macro-based additions, use
+the :code:`get_schema_with_macros` function:
+
+.. code:: python
+
+    from graphql_compiler.macros import get_schema_with_macros
+
+    graphql_schema = get_schema_with_macros(macro_registry)
```

Comment:
I think that you should show at one point what an GraphQL object type with macros looks like.  (Not the entire schema because that would be too much text).

Reply:
Good point. Unfortunately, currently the `get_schema_with_macros()` function is slightly broken because of #521 so the macro edge directive won't show up, and the example will either be lying or be not that useful.

Perhaps in a follow-up PR, I can extract a new function (e.g. `get_schema_ast_with_macros()`) that is called by `get_schema_with_macros()`. This new function would return a schema AST, which when printed would contain the `@macro_edge` directives in the appropriate places and would therefore be suitable for an example. I've made #588 for this purpose.

Reply:
Even if the field is not marked by a macro edge directive, just seeing how a new field gets added to the schema would help the user understand macros. There is also an `:emphasize-lines:` option from sphinx that would fit in nicely here so it might be worthwhile to add a TODO for that if you proceed with this approach.

I would rather not talk about the ast because we don't talk about the ast anywhere else in the documentation and some users might not even know what an ast is. Also asts are not necessarily the easiest concept to understand/explain.

Reply:
For what it's worth, based on chatting with a number of GraphQL users at a meetup a couple of weeks ago, it seems that ~nobody reads GraphQL schemas manually. Based on that, while I agree that more examples would be helpful, I don't think this is a high value-add relative to the time we'd spend on it.

In my opinion, it might be better to fix the schema-generation API to have an AST-based option, potentially move to ReadTheDocs in the meantime, and then update the docs to something even better. I think this is good enough in the meantime.